File: MAP/lab7/lab7/Models/FileHandlers/ReadFile.cs
using System;
using System.IO;

namespace lab7.Models.FileHandlers
{
    public class ReadFile : IStatement
    {
        private IExpression expID;
        private string newVarName;

        public ReadFile(IExpression expId, string newVarName)
        {
            expID = expId;
            this.newVarName = newVarName;
        }


        public PrgState execute(PrgState state)
        {
            int id = expID.eval(state.Dict);
            StreamReader sw = state.FileTable.get(id).Sw;

            string line = sw.ReadLine();
            if(line == "")
                throw new SystemException("Can't read the line");
            
            
            int value = int.Parse(line);
            
            state.Dict.setValue(newVarName, value);

            return state;


        }

        public override string ToString()
        {
            return "ReadFile(" + newVarName + ")";
        }
    }
}

File: MAP/lab8 - ProcTable/src/models/fileHandling/ReadFile.java
package models.fileHandling;

import exceptions.FileException;
import models.PrgState;
import models.expression.IExpression;
import models.statement.Statement;
import utils.IDictionary;
import utils.IHeap;

import java.io.BufferedReader;
import java.io.IOException;

public class ReadFile implements Statement {

    IExpression exp;
    String varName;

    public ReadFile(IExpression exp, String varName) {
        this.exp = exp;
        this.varName = varName;
    }


    @Override
    public PrgState execute(PrgState state) {

        IDictionary<String, Integer> dict = state.peekStackSymTab();
        IFileTable<Integer, FileData> fileTable = state.getFileTable();
        IHeap<Integer, Integer> heap = state.getHeap();
        int id = exp.eval(dict, heap);



        if(!fileTable.contains(id))
            throw new FileException("This ID is not in FileTable");

        BufferedReader reader = fileTable.get(id).getReader();

        String line = null;
        try {
            line = reader.readLine();
        } catch (IOException e) {
            throw new FileException("Can't read line");
        }

        int value = 0;
        if(line != null)
            value = Integer.parseInt(line);

        dict.setValue(varName, value);

        return null;
    }

    @Override
    public String toString(){
        return "read("+varName + ", " + exp+")";
    }

}
